src/Fixer/LanguageConstruct/NullableTypeDeclarationFixer.php

```diff
@@ -0,0 +1,326 @@
+    {
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder('syntax', 'Whether to use question mark (`?`) or explicit `null` union for nullable type.'))
+                ->setAllowedValues([self::OPTION_USE_EXPLICIT_NULL, self::OPTION_USE_QUESTION_MARK])
+                ->setDefault(self::OPTION_USE_QUESTION_MARK)
+                ->getOption(),
+        ]);
+    }
+
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
+    {
+        $functionsAnalyzer = new FunctionsAnalyzer();
+
+        foreach (array_reverse($this->getElements($tokens), true) as $index => $type) {
+            if ('property' === $type) {
+                $this->normalizePropertyType($tokens, $index);
+
+                continue;
+            }
+
+            // fix return before arguments
+            $this->normalizeMethodReturnType($functionsAnalyzer, $tokens, $index);
+            $this->normalizeMethodArgumentType($functionsAnalyzer, $tokens, $index);
+        }
+    }
+
+    /**
+     * @return array<int, string>
+     *
+     * @phpstan-return array<int, 'method'|'property'>
+     */
+    private function getElements(Tokens $tokens): array
+    {
+        $tokensAnalyzer = new TokensAnalyzer($tokens);
+
+        $elements = array_map(
+            static fn (array $element): string => $element['type'],
+            array_filter(
+                $tokensAnalyzer->getClassyElements(),
+                static fn (array $element): bool => \in_array($element['type'], ['method', 'property'], true)
+            )
+        );
+
+        foreach ($tokens as $index => $token) {
+            if (
+                $token->isGivenKind(T_FN)
+                || ($token->isGivenKind(T_FUNCTION) && !isset($elements[$index]))
+            ) {
+                $elements[$index] = 'method';
+            }
+        }
+
+        return $elements;
+    }
+
+    private function collectTypeAnalysis(Tokens $tokens, int $startIndex, int $endIndex): ?TypeAnalysis
+    {
+        $type = '';
+        $typeStartIndex = $tokens->getNextMeaningfulToken($startIndex);
+        $typeEndIndex = $typeStartIndex;
+
+        for ($i = $typeStartIndex; $i < $endIndex; ++$i) {
+            if ($tokens[$i]->isWhitespace() || $tokens[$i]->isComment()) {
+                continue;
+            }
+
+            $type .= $tokens[$i]->getContent();
+            $typeEndIndex = $i;
+        }
+
+        return '' !== $type ? new TypeAnalysis($type, $typeStartIndex, $typeEndIndex) : null;
+    }
+
+    private function isTypeNormalizable(TypeAnalysis $typeAnalysis): bool
+    {
+        if ($typeAnalysis->isNullable()) {
+            return true;
+        }
+
+        $type = $typeAnalysis->getName();
+
+        if (str_contains($type, '&')) {
+            return false; // skip DNF types
+        }
+
+        return str_contains($type, '|')
+            && 1 === substr_count($type, '|')
+            && 1 === Preg::match('/(?:\|null$|^null\|)/i', $type);
```

Comment:
you should work on type tree instead of string, I am not sure if native type is currently parsed, but you can use phpdoc TypeExpression parser which will always parse native type, as native type grammar will be always a subset of rich phpdoc grammar

Reply:
I think this is a future enhancement. Since I am using the `TypeAnalysis` class, we can adapt it to return a `TypeExpression` object of the types it has.

Reply:
It won't parse `(x|y)`, it is a ticking bomb to me.

Reply:
Maybe I did not get what you mean, but this does not need to parse that. It just needs to know if there is a sticking `null` at either end of unioned type. So in `null|x` and `x|null` it does not need to know what the other type is.

Wait, if you mean the wrapping parenthesis, since this deals with actual types (not in PHPDocs) and there's no support yet of conjunctive normal form type definitions, there's no need to worry about that yet. Well, if support comes later on then we can add complex parsing. For now, [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

Reply:
`(x|y)` seems to be not allowed: https://3v4l.org/Fv00q/rfc#vgit.master_jit

but https://3v4l.org/nebaI/rfc#vgit.master_jit will fail for example ;-)

I simply belive you should not use string manipulation here, in past, such approach was generating a lot of edge cases hidden bugs.

Reply:
FYI: I also would like to use approach based on type collections rather that string and regexes, but it's good enough for now I think.